### PR TITLE
HHH-11520 HHH-11517 Cloud friendliness and connection to DBs not yet present

### PR DESCRIPTION
**Pull request by emmanuelbernard**
The goal is to help people deploying Hibernate ORM apps when their database might be started after the Hibernate app. Cloud providers might offer some solution but this work helps users even if cloud providers are not very cooperative.

See http://in.relation.to/2017/02/16/hibernate-connections-cloud/ for more context.

## What has been done

HHH-11517:
I've fixed a few issues in particular from the built-in connection pool that was not reactive well to the DB not being present during configuration time. This means the work on `JDBCEnvironmentInitiator` falling back to default if the connection is not available now works in more situations

I've added the ability to delay some bootstrap operations needing connections to the first time a valid connection is retrieved.
I've added the ability for the schema generation tools to run as delayed operation.

## What is planned

More at https://hibernate.atlassian.net/browse/HHH-11520

Allow the connection provider to circuit break when the database is not present or verify that all connection pools offer such options.

Consider enforcing specific connection pool options to be friendly toward booting with no database available (e.g. Hikari) when the hibernate.tolerate_database_not_present option is set to true.

Create a specific how-to documentation explaining how to work in environments where the DB might not be present at boot time. Update the reference doc as well.

Consider adding an API for a user or health check implement to know our current DB status (delayed vs ready)

Explore ability to extract JDBC environment information from DB connection in a delayed fashion: that's going to be the hardest most likely and we could start without this option.

## Feedback on general approach

What I want to know is whether you think I'm going in the right direction and whether I'm doing idiomatic Hibernate ORM development (it's been a while so I might be misusing some APIs or using wrong patterns.

Don't integrate it yet as I want to do some of what is planned but comments and feedback are welcome.

Feedback on the planned tasks would be useful too.

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
+			}
+		}
+		catch (Exception e) {
+			// we don't care if there is a failure
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: I would log the exception, with a Map<String,Runnable> instead of a List<Runnable> where the key is the name of the operation we could log also which operation failed

Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: stupid comment, forget about it

File: hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java [hunk 2]
```diff
@@ -1646,4 +1646,20 @@
...
+	 * There are a few things con consider:
+	 * <ul>
+	 *     <li>schema generation will be delayed until the first successful connection</li>
+	 *     <li>you must set {@link AvailableSettings#DIALECT} explicitly as we cannot guess the database if it's not there</li>
+	 *     <li>some JDBC specific settings set based on the JDBC metadata will not be set and best effort defaults will apply
+	 *     if the database is not present at Hibernate boot time</li>
+	 * </ul>
+	 * @since 6.0
+	 */
+	String TOLERATE_DATABASE_NOT_PRESENT = "hibernate.tolerate_database_not_present";
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/cfg/AvailableSettings.java]: The property name is a bit puzzling, it might suggest people that Hibernate could "work" even in absence of the database.

Maybe invert the boolean and make it "fail_fast_on_missing_database" ?

Or  "wait_for_database" ?

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java [hunk 3]
```diff
@@ -99,17 +101,22 @@ public ConnectionProvider initiateService(Map configurationValues, ServiceRegist
...
 		}
 
+		boolean tolerateAbsentDB = ConfigurationHelper.getBoolean(
+				AvailableSettings.TOLERATE_DATABASE_NOT_PRESENT,
+				configurationValues,
+				false
+		);
 		final StrategySelector strategySelector = registry.getService( StrategySelector.class );
 		final Object explicitSetting = configurationValues.get( AvailableSettings.CONNECTION_PROVIDER );
 		if ( explicitSetting != null ) {
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/ConnectionProviderInitiator.java]: extract this whole block, to have a single wrapWithDelayer() invocation?

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+				}
+				finally {
+					lock.unlock();
+				}
+			}
+		}
+
+		return connection;
+	}
+
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: What concerns me is the schema creation or validation may require a lot of time to be performed and when delayed it may be triggered by session.beginTransaction() call.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	private void tryAndExecuteOperations() {
+		// It would be more elegant to refactor the operation lock from getConnection is a shared method
+		// but I prefer the JVM to have more chances to keep getConnection() inlined
+		try {
+			Connection c = getConnection();
+			closeConnection( c );
+			if (operations == null) {
+				// operation execution has succeeded, but subsequent ones might happen
+				// remember, we are not multi-threaded here as we are bootstrapping Hibernate ORM
+				operations = new ArrayList<>();
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: Something is dodgy here. Why would you think this `operations` could be null at this point?

This method is invoked only by _addOperation_, which just used the `operations` list. Should `addOperation` rather guard against null, and possibly initialize it?

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+			if ( ! lock.isHeldByCurrentThread() ) {
+				lock.lock();
+				try {
+					if ( operations != null ) {
+						// get a service registry bypassing the DelayerConnectionProvider for the action calls
+						//ServiceRegistry sr = buildServiceRegistryDelegator();
+						operations.forEach( Runnable::run );
+						// if an operation fails, the list is not cleared but operations are
+						// catastrophic events and further use should not happen
+						operations = null;
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: It seems this is the only point which can eventually null-out the `operations` field.
Could you maybe?
 - elaborate on why it's being nulled? I'm not understanding the comment.
 - maybe avoid it? That way you could make the `operations` field final and simplify this code a lot.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+		}
+		catch (Exception e) {
+			throw e;
+		}
+
+		// protect operation execution behind a double checked lock
+		// we don't want them to be applied multiple times
+		if ( operations != null ) {
+			// if lock is already held, we are reentering and acquiring a connection for the an operation execution
+			// so we need to ignore operations here
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: Is there a strong guarantee that all "reentrant" operations are executed by the same thread?
Maybe a different kind of switch would make this code future-proof to concurrent initialization.

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+			Configurable configurableConnectionProvider = (Configurable) connectionProvider;
+			configurableConnectionProvider.configure( configurationValues );
+		}
+	}
+
+	@Override
+	public Connection getConnection() throws SQLException {
+		Connection connection;
+		try {
+			connection = connectionProvider.getConnection();
```
Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: I'm wondering - not sure at all - is it possible that the `getConnection()` operation could be affected by the pending operations?

Maybe to play it safe I'd move the `getConnection` invocation to after the following block of code, essentially after the operations have been applied.

Comment (Sanne) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/connections/internal/DelayerConnectionProvider.java]: A more concrete example: suppose this is invoked in the scope of a transaction for thread A. Another thread B just got the connection, and is now going to execute DDL operations. Thread A is waiting for those to complete, but when it gets the green light it might not have the tables yet, as its transaction started before they were created.

**Comment by sebersole** [comment]
First, can we draw some boundaries here?  E.g. do we handle the connections going away after we have previously had access?  Personally I think this should not be supported; or more correctly I think our current behavior is probably correct for this use case.

That leaves just this bootstrap concern - the connections not *yet* being available initially...  I think we probably all agree that this one is a worthwhile goal.

Regarding TheRightWay(tm), IMO that comes down to a number of things.  The big one is our view of the database, especially in regards to db-object names that get determined partially from information gleaned from the Connection.  ATM this happens up front (even in your proposal still). This effectively means that we cannot properly perform the `Metamodel` -> "run-time model" (persisters) conversion until that time also.  Basically we'd have to delay the building the entire SF.  Also, we'd need to be able to delay creation of other services as well[1].

To illustrate this problem, create a model that tries to define a column whose name is 50 characters long.  For Oracle we know we'd have to trim that to 30 chars or less.  Those truncated names are what get incorporated into the run-time model.  So what you'd see is the run-time model having incorporated the incorrect (50 char long) names.  So that SF would forever be useless[2].

Honestly, this is a great illustration of why it is better to model a Column in the model rather than column name Strings (and there are many additional reasons as well); if all uses have reference to that Column I can easily change/define that in one place - updating the name on that Column reference it "propagate" to all uses (FKs, etc) because they hold the Column reference.  With the current Strings approach I'd have to keep track of all the locations where that column is used and update all of them

I know the connection is not immediately obvious, but IMO a huge prerequisite for doing this *properly* is some of the work happening for 6.0 as well as the "jandex-binding" work which is planned to be the base for 7.0.  Why?  The pertinent 6.0 change is in fact definition of such Column objects in the run-time model in addition to a central "Database" reference (as opposed to today where we have to walk the entire run-time model and collect/interpret the db-objects.  The 7.0 changes would have an effect in terms of not applying `PhysicalNamingStrategy` conversions (truncation like above, etc) until we are in the process of building that run-time model (at which point the database would be required to be available).

So that's long term.  What do we do in the short term?  I think the first question is which version to base this on.  These proposed changes apply to 5.2.  Personally I am not sure that is doable in 5.2 for the reasons already mentioned.  And targeting 6.0 is not really "short term" when we look at containers using that version[3]

[1] Perhaps this is better handled more generically as a `Service`/`ServiceRegistry` feature.  Something like letting `ServiceA` declare that it should not be started until `ServiceB` has successfully started.  The immediate question there is how we define/expose the idea of "successfully started".  Perhaps a `Future`?  Anyway, the reason I bring this up is that it is not just the `ConnectionProviderService`; this affects `JdbcServices` (and its collected services), etc.

[2] Note that your tests only build the SF.  Your change wrt allowing a delayed Connection access only works wrt building the SF.  You'd have to try to use it to see the rest of the breakages like the ones I mentioned.

[3] Unless we consider our `hibernate-modules` as fulfilling that[4].

[4]  BTW, can we please rename that to `hibernate-wildfly-modules`?  IMO `hibernate-modules` is quite unclear.

**Comment by sebersole** [comment]
To clarify my earlier comment agreeing with Andrea...

That is in regards to the `#wrapWithDelayer` method.  ATM if this "toleration" setting is enabled you just immediately wrap the real ConnectionProvider.  The problem here is that all calls to the ConnectionProvider will now go through 2 methods for every method called (the call on the delayer delegate and the delegation) - so we ought to be as defensive as possible in regards to *not* using this delayer IMO.  So in your `#wrapWithDelayer` method we should ask the ConnectionProvider whether it is "ready for use" - only if not should we look at the "tolerate" setting.  In other words, instead of:

````
    private ConnectionProvider wrapWithDelayer(
            ConnectionProvider connectionProvider, 
            boolean tolerateDatabaseNotPresent) {
                return tolerateDatabaseNotPresent 
                        ? new DelayerConnectionProvider( connectionProvider ) 
                        : connectionProvider;
    }
````

do:
````
    private ConnectionProvider wrapWithDelayer(
            ConnectionProvider connectionProvider, 
            boolean tolerateDatabaseNotPresent) {
                ConnectionProvider providerToUse = connectionProvider;
                if ( tolerateDatabaseNotPresent ) {
                    try {
                        // try to get Connection and use it
                    }
                    catch (SqlException ignore) {
                        providerToUse = new DelayerConnectionProvider( connectionProvider );
                    }
                }
            }
        }
    }
````

Yet another option is to include this logic in our ConnectionProvider impls.

**Comment by Sanne** [comment]
Why would you prefer that?  I'd prefer to encapsulate the "ignore SQL and other errors" logic in a single place.

**Comment by Sanne** [comment]
> [4] BTW, can we please rename that to hibernate-wildfly-modules? IMO hibernate-modules is quite unclear.

Good idea - > https://hibernate.atlassian.net/browse/HHH-11528

> 2] Note that your tests only build the SF. Your change wrt allowing a delayed Connection access only works wrt building the SF. You'd have to try to use it to see the rest of the breakages like the ones I mentioned.

I'm lost here :) Which other breackages? Clearly the intent here is just to delay boot strap, not solve other problems like the ones caused by connections being lost after boot.

**Comment by dreab8** [comment]
I think we have also to understand what can be delayed and expecially till when, I think operations like schema validation or schema creation should be performed before a session is created otherwise are meaningless.

**Comment by sebersole** [comment]
> I'm lost here :) Which other breackages? Clearly the intent here is just to delay boot strap, not solve other problems like the ones caused by connections being lost after boot.

I missed that the test is explicitly setting the Dialect.

**Comment by sebersole** [comment]
> Why would you prefer that? I'd prefer to encapsulate the "ignore SQL and other errors" logic in a single place.
No idea what this is in reference to.  You mean you'd prefer to just always wrap the ConnectionProvider when this setting is enabled even when the connections are available?  That seems like an obvious performance problem to me...

**Comment by Sanne** [comment]
> No idea what this is in reference to. You mean you'd prefer to just always wrap the ConnectionProvider when this setting is enabled even when the connections are available? That seems like an obvious performance problem to me...

Right, sorry I was referring to your comments regarding #wrapWithDelayer

Ok, I wasn't sure you were referring to performance. I am not sure it would be a problem, the check in the wrapper can be made to be extremely cheap, possibly negligible.
You might be right, but it seems premature to worry about it w/o runtime data. Design wise I would prefer to wrap it unconditionally (if the feature is enabled), so that then we have a good spot to evolve the approach also for implementing the follow-ups, such as handling the temporary connection at runtime (after bootstrap).

**Comment by sebersole** [comment]
I hadn't thought of the "handling the temporary connection .. after bootstrap" because as you said "the intent here is just to delay boot strap, not solve other problems like the ones caused by connections being lost after boot".  So I took it on that level ;)

Also, regarding "encapsulate the "ignore SQL and other errors" logic in a single place".  But that's a problem.  Encapsulate them to what level/scope?  Only to acquiring a Connection?  Ok, simply ignoring the Connection being unavailble is ever only going to be useful at boot-time anyway.  At runtime trying to acquire the Connection is always the precursor to Hibernate trying to execute some SQL.  In this case we most definitely want the unavailable Connection to be an error.

Also, unless you intend to wrap all JDBC operations in a call to this "encapsulated" code to apply the unified handling then it can only be scoped at the Connection level as discussed above.

**Comment by sebersole** [comment]
TLDR; I do not think we want to do anything here

To hopefully come to some conclusion here, let's look at the various situations...

1) Connections are available when the SF is bootstrapped, but go away later.  I do not think we should do any special behavior here.  This is a use case for HR in my opinion

2) Connections are not available when the SF is bootstrapped.  Here we have 2 concerns:

2.a) Access to DatabaseMetaData to apply implicit configuration of the SF.  Again, I absolutely think this is something we do not want to handle.  If the Connection is not available during bootstrap, we should continue to expect that the user has given us proper configuration.  
The way I look at this... this is the responsibility of the thing boostrapping the SF (JPA container, OSGi container, etc)

2.b) Delay auto schema tooling.  This is something we could consider, but for the same reason as (2.a) I'm not sure we should be.  IMO this should fall under "the responsibility of the container"

**Comment by sebersole** [comment]
We all agreed that ORM is not the place to handle these concerns
